Add rgw_enable_apis option to ObjectStore CRD

arttor opened the pull request:
Closes #15043

- Added [rgw_enable_apis](https://docs.ceph.com/en/reef/radosgw/config-ref/#confval-rgw_enable_apis) option to `CephObjectStore` CRD.
- Moved existing `rgw_enable_apis` config from mon db to CLI flags to restart RGW on update.
- Adjusted RGW probes to work when only Swift or only admin-ops api enabled.
- docs for multi-instance rgw deployment

arttor commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1741,6 +1741,14 @@ type EndpointAddress struct {
+	// This option overrides S3.Enabled value.
+	// +optional
+	// +nullable
+	EnableAPIs []ObjectStoreAPI `json:"enableAPIs,omitempty"`
```
Does it make sense instead to use something like this:
```go
type ObjectStoreSpec struct{
   Config map[string]string
  ...
```
to be able to set any rgw options?

BlaineEXE replied:
I have had some thoughts to enable something like this. I think it would reasonably be its own separate PR, but I am interested in exploring this. LMK what you think

arttor replied:
good idea! in this case it should definetely go in a separate PR.

arttor commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1752,8 +1760,12 @@ type ProtocolSpec struct {
+
 // S3Spec represents Ceph Object Store specification for the S3 API
 type S3Spec struct {
+	// Deprecated: use protocol.enableAPIs instead.
```
Is it ok to deprecate?

BlaineEXE replied:
I think swift support is marked experimental, so this can be deprecated without concern. However, I would think maybe we don't backport that to 1.15 if so. 

This would be a good discussion to include @kayrus and @jklippel in I think.

BlaineEXE commented on deploy/examples/object-multi-instance-test.yaml:
```diff
@@ -0,0 +1,107 @@
+      - s3
+      - s3website
+      - sts
+      - iam
```
I have been wondering about this for a while: is `iam` permission something users should have, or admin? This may be a matter of opinion and user intent as well. No change requested here, but something I'm curious about.

arttor replied:
Good question. I hope that it will work both ways but ceph docs says:

> This REST API is served by the same HTTP endpoint as the Ceph Object Gateway S3 API.

BlaineEXE commented on deploy/examples/object-multi-instance-test.yaml:
```diff
@@ -0,0 +1,107 @@
+    replicated:
+      size: 1
+      requireSafeReplicaSize: false
+    parameters:
+      compression_mode: none
```
Minor nit: For targeted examples, it can be helpful for readability to remove configs that aren't relevant. I think `compression_mode`'s default config is `none`, so these lines can be removed.

BlaineEXE commented on deploy/examples/object-multi-instance-test.yaml:
```diff
@@ -0,0 +1,107 @@
+apiVersion: ceph.rook.io/v1
+kind: CephObjectRealm
+metadata:
+  name: my-store
```
Given that `my-store` is the name of default `object*.yaml` files, I think we should change this to ensure no conflicts. `multi-instance` would be fine, or really anything else.

BlaineEXE commented on deploy/examples/object-multi-instance-test.yaml:
```diff
@@ -0,0 +1,107 @@
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStore
+metadata:
+  name: store-admin
```
If my plan is right, users can consume this object store by pointing child resources to the admin. Since that is a confusing change from other object stores, I would suggest that we add a CephObjectStoreUser to this example to illustrate how to use those with this setup. That resource will also make it easy to ensure the sample is working, by checking whether the COSUser status.

IMO, it might also be reasonable to add an OBC StorageClass (not necessarily a whole OBC), also for illustration purposes, but I'll let you decide on that.

arttor replied:
I've added `CephObjectStoreUser` to this example and also to docs. However, when i tested it locally i noticed that generated user secret also containts `Endpoint`. And in this case it will point to admin RGW but if i got it right created user should be used for s3 api.
```shell
k get secrets rook-ceph-object-user-store-admin-multi-instance-user -o jsonpath={.data.Endpoint} | base64 -d
http://rook-ceph-rgw-store-admin.rook-ceph.svc:80
```
Is Endpoint field used?

BlaineEXE replied:
Sorry for the delay. Yes, the endpoint is used, or at least it's meant to be usable. Now that you've mentioned that behavior, it is obvious to me that something else is needed.

To make things integrate well, we will need to parameterize something else in the object store. My suggestion would be that we keep this PR as-is, while keeping in mind that the endpoints won't be correct for CephObStoreUser or OBCs with just this change. I would prefer to follow up with design/discussion about how to proceed for better integration. I think this change is reasonable standalone.

While I don't love the idea of having a configuration to specify the admin ops endpoint, I think that probably ends up being the simplest for users to understand, and for us to document. An alternative is to consider reusing the `advertiseEndpoint` as the inverse, but that seems a little backwards logically, and I think that would make it harder to use and explain.

BlaineEXE commented on pkg/operator/ceph/object/spec.go:
```diff
@@ -489,13 +499,50 @@ func (c *clusterConfig) defaultReadinessProbe() (*v1.Probe, error) {
+		prefix += "info"
+		return prefix, 0
+	}
+	if slices.Contains(enabledAPIs, "admin") {
+		// using admin-ops api for probe
+		// check that admin ops info responds with 403 - Forbidden because all admin endpoinds secured
+		return "/admin/info", 403
+	}
```
I like what you're doing here. I think I want to get @cbodley 's thoughts also. If we get startup/readiness probes wrong in corner cases, it can prevent them from starting up or from being accessible even if they could be accessed. I'd like to avoid that as much as possible.

Casey, is there a recommended way to check RGW health for an RGW that only has the `admin` API enabled (no S3/swift or other user endpoints)?

BlaineEXE commented on pkg/operator/ceph/object/spec.go:
```diff
@@ -489,13 +499,50 @@ func (c *clusterConfig) defaultReadinessProbe() (*v1.Probe, error) {
+		// check that admin ops info responds with 403 - Forbidden because all admin endpoinds secured
+		return "/admin/info", 403
+	}
+	logger.Warningf("no suitable for probe api enabled %v. Fallback to s3 probe", enabledAPIs)
```
Is this better wording?

```suggestion
	logger.Warningf("no suitable api for probe enabled %v. Fallback to s3 probe", enabledAPIs)
```

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1752,8 +1760,12 @@ type ProtocolSpec struct {
 	Swift *SwiftSpec `json:"swift"`
 }
 
+// +kubebuilder:validation:Enum=s3;s3website;swift;swift_auth;admin;sts;iam;notifications
+type ObjectStoreAPI string
```
Nice!

BlaineEXE reviewed (changes requested):
Overall, this looks great as usual. The biggest concern I have is with the startup/readiness probe.

My concern with these probes is not specifically your implementation but with usage of the probes at all. Kubernetes warns against using probes when they aren't accurate, because it can lead to unnecessary API calls and instability of the underlying application due to constant daemon resets.

To be honest, I really question whether Rook's usage of probes is actually beneficial for Ceph and/or users. When there are errors in probes, it can cause startup/availability errors, and that seems worse to me than a pod going into a hung state or CLBO and needing a manual restart. And, frankly, Rook has gotten probes wrong more often than right since they were first (IMO hastily) implemented years and years ago.

Obviously, we still have them for S3, but you've observed that there are corner cases where it breaks. E.g., a user reported an issue when using host networking.

I don't have much of an opinion either way on whether to continue doing probes for the swift API.

But I'd like to suggest and get your opinion on whether it might be better to not use a probe when the s3 and swift are both disabled. This would include the case where it's just the admin API, but it would also include weird corner cases like if someone just enables the `sts` API or something. IMO, the best default for probes is to not have one. This would also make the `OKCode` complexity unnecessary.

BlaineEXE commented on pkg/operator/ceph/object/spec.go:
```diff
@@ -489,13 +499,50 @@ func (c *clusterConfig) defaultReadinessProbe() (*v1.Probe, error) {
+		// check that admin ops info responds with 403 - Forbidden because all admin endpoinds secured
+		return "/admin/info", 403
+	}
+	logger.Warningf("no suitable for probe api enabled %v. Fallback to s3 probe", enabledAPIs)
+	return "", 0
```
I mentioned this in my review, but I want to flag it on the code as well. IMO, the best fallback when we don't know what to do is to turn the probe off instead so that we don't risk causing daemon crash loop issues without clear purpose.

arttor commented:
@BlaineEXE can you please take  a look again. I fixed probe, added docs and added User to examples. 
However, there might be a problem with user and admin-only RGW because created secrets also contains endpoint.
